### PR TITLE
Update OTA fields for model version

### PR DESCRIPTION
**Pull request by DenisRybas**
Summary:
All OTA fields are always editable if not set initially by a Vendor Account with the same VID as the one who initially created the Model Version
Link to ticket: https://github.com/zigbee-alliance/distributed-compliance-ledger/issues/466

### CODE REVIEW

File: x/model/keeper/msg_server_model_version.go [hunk 1]
```diff
@@ -109,15 +104,35 @@ func (k msgServer) UpdateModelVersion(goCtx context.Context, msg *types.MsgUpdat
 		return nil, types.NewErrMaxSVLessThanMinSV(modelVersion.MinApplicableSoftwareVersion, msg.MaxApplicableSoftwareVersion)
 	}
 
-	// update existing model version value only if corresponding value in MsgUpdate is not empty
+	if msg.OtaUrl != "" && modelVersion.OtaUrl != "" && msg.OtaUrl != modelVersion.OtaUrl {
```
Comment (ashcherbakov) [inline x/model/keeper/msg_server_model_version.go:107]: The check seems wrong. Why do we need it? OtaURL should be always mutable without any restrictions.

File: x/model/keeper/msg_server_model_version.go [hunk 2]
```diff
@@ -109,15 +104,35 @@ func (k msgServer) UpdateModelVersion(goCtx context.Context, msg *types.MsgUpdat
...
+		return nil, types.NewErrOtaURLCannotBeSet(modelVersion.Vid, modelVersion.Pid, modelVersion.SoftwareVersion)
+	}
 
-	// SoftwareVersionValid flag is updated in any case. So pass the existing value to keep it unchanged.
-	modelVersion.SoftwareVersionValid = msg.SoftwareVersionValid
+	if msg.OtaFileSize != 0 && modelVersion.OtaFileSize != 0 && msg.OtaFileSize != modelVersion.OtaFileSize {
+		return nil, types.NewErrOtaFileSizeCannotBeSet(modelVersion.Vid, modelVersion.Pid, modelVersion.SoftwareVersion)
+	}
+
+	if msg.OtaChecksum != "" && modelVersion.OtaChecksum != "" && msg.OtaChecksum != modelVersion.OtaChecksum {
```
Comment (ashcherbakov) [inline x/model/keeper/msg_server_model_version.go:115]: The check seems wrong. Why do we need it? msg.OtaChecksum should be always mutable without any restrictions.

File: x/model/types/errors.go [hunk 3]
```diff
@@ -24,7 +24,7 @@ var (
 	ErrMaxSVLessThanMinSV            = sdkerrors.Register(ModuleName, 521, "max software version less than min software version")
 	ErrLsfRevisionIsNotValid         = sdkerrors.Register(ModuleName, 522, "LsfRevision should monotonically increase by 1")
 	ErrLsfRevisionIsNotAllowed       = sdkerrors.Register(ModuleName, 523, "LsfRevision is not allowed if LsfURL is not present")
-	ErrModelVersionDeletionCertified = sdkerrors.Register(ModuleName, 524, "model version has a compliance record and can not be deleted")
+	ErrModelVersionDeletionCertified = sdkerrors.Register(ModuleName, 524, "model version certified and can not be deleted")
```
Comment (ashcherbakov) [inline x/model/types/errors.go:27]: Why is it changed?

File: x/model/types/errors.go [hunk 4]
```diff
@@ -102,7 +102,7 @@ func NewErrModelVersionAlreadyExists(vid interface{}, pid interface{}, softwareV
 func NewErrOtaURLCannotBeSet(vid interface{}, pid interface{}, softwareVersion interface{}) error {
```
Comment (ashcherbakov) [inline x/model/types/errors.go:102]: It seems this function is not used now and can be deleted.

File: x/model/handler_test.go [hunk 5]
```diff
@@ -698,6 +698,9 @@ func TestHandler_UpdateModelVersion(t *testing.T) {
 
 	// try update not present model version
 	msgUpdateModelVersion := NewMsgUpdateModelVersion(setup.Vendor)
+	msgUpdateModelVersion.OtaUrl = ""
```
Comment (ashcherbakov) [inline x/model/handler_test.go:701]: Why do we need to set these values to default in so many places? Aren't they default by default? If not, then it cane sent to default in `NewMsgUpdateModelVersion` method (in one place).

File: x/model/handler_test.go [hunk 6]
```diff
@@ -727,7 +730,7 @@ func TestHandler_UpdateModelVersion(t *testing.T) {
 	require.Equal(t, receivedModelVersion.SoftwareVersion, msgUpdateModelVersion.SoftwareVersion)
 
 	require.Equal(t, receivedModelVersion.SoftwareVersionValid, msgUpdateModelVersion.SoftwareVersionValid)
-	require.Equal(t, receivedModelVersion.OtaUrl, msgUpdateModelVersion.OtaUrl)
+	require.Equal(t, receivedModelVersion.OtaUrl, msgCreateModelVersion.OtaUrl)
```
Comment (ashcherbakov) [inline x/model/handler_test.go:733]: Please check that other OTA fields are also not updated

**Review by ashcherbakov — CHANGES_REQUESTED** [review]
1) Please update documentation 
2) Please extend the UpdateModelVersion tests for gRPC/REST with OTA fields.

File: x/model/handler_test.go (hunk 6 above)
Comment (DenisRybas) [reply x/model/handler_test.go]: Checked

File: x/model/handler_test.go (hunk 5 above)
Comment (DenisRybas) [reply x/model/handler_test.go]: Fixed

File: x/model/types/errors.go (hunk 4 above)
Comment (DenisRybas) [reply x/model/types/errors.go]: Deleted

File: x/model/types/errors.go (hunk 3 above)
Comment (DenisRybas) [reply x/model/types/errors.go]: Fixed

File: x/model/handler_test.go [hunk 7]
```diff
@@ -727,16 +728,18 @@ func TestHandler_UpdateModelVersion(t *testing.T) {
...
 	require.Equal(t, receivedModelVersion.MinApplicableSoftwareVersion, msgUpdateModelVersion.MinApplicableSoftwareVersion)
 	require.Equal(t, receivedModelVersion.MaxApplicableSoftwareVersion, msgUpdateModelVersion.MaxApplicableSoftwareVersion)
 	require.Equal(t, receivedModelVersion.ReleaseNotesUrl, msgUpdateModelVersion.ReleaseNotesUrl)
 
 	require.Equal(t, receivedModelVersion.SoftwareVersionString, msgCreateModelVersion.SoftwareVersionString)
 	require.Equal(t, receivedModelVersion.CdVersionNumber, msgCreateModelVersion.CdVersionNumber)
 	require.Equal(t, receivedModelVersion.FirmwareInformation, msgCreateModelVersion.FirmwareInformation)
-	require.Equal(t, receivedModelVersion.OtaFileSize, msgCreateModelVersion.OtaFileSize)
-	require.Equal(t, receivedModelVersion.OtaChecksum, msgCreateModelVersion.OtaChecksum)
+	require.Equal(t, receivedModelVersion.OtaFileSize, msgCreateModelVersion.OtaFileSize+1)
```
Comment (ashcherbakov) [inline x/model/handler_test.go:741]: It seems we've already checked it above.

File: x/model/handler_test.go [hunk 8]
```diff
@@ -819,6 +822,9 @@ func TestHandler_UpdateOnlyMinApplicableSoftwareVersion(t *testing.T) {
 
 	// try to update only min version to a value greater than stored max version
 	msgUpdateModelVersion := NewMsgUpdateModelVersion(setup.Vendor)
+
+	msgUpdateModelVersion.OtaChecksum = ""
```
Comment (ashcherbakov) [inline x/model/handler_test.go:826]: Why do we need to set it here?

File: x/model/handler_test.go [hunk 9]
```diff
@@ -828,6 +834,9 @@ func TestHandler_UpdateOnlyMinApplicableSoftwareVersion(t *testing.T) {
 
 	// try to update only min version to a value less than stored max version
 	msgUpdateModelVersion = NewMsgUpdateModelVersion(setup.Vendor)
+
+	msgUpdateModelVersion.OtaChecksum = ""
```
Comment (ashcherbakov) [inline x/model/handler_test.go:838]: Why do we need to set it here?

File: x/model/handler_test.go [hunk 10]
```diff
@@ -849,6 +858,9 @@ func TestHandler_UpdateOnlyMinApplicableSoftwareVersion(t *testing.T) {
 
 	// try to update only min version to a value equal to stored max version
 	msgUpdateModelVersion = NewMsgUpdateModelVersion(setup.Vendor)
+
+	msgUpdateModelVersion.OtaChecksum = ""
```
Comment (ashcherbakov) [inline x/model/handler_test.go:862]: Why do we need to set it here?

File: x/model/handler_test.go [hunk 11]
```diff
@@ -895,6 +910,9 @@ func TestHandler_UpdateOnlyMaxApplicableSoftwareVersion(t *testing.T) {
 
 	// try to update only max version to a value greater than stored min version
 	msgUpdateModelVersion = NewMsgUpdateModelVersion(setup.Vendor)
+
+	msgUpdateModelVersion.OtaChecksum = ""
```
Comment (ashcherbakov) [inline x/model/handler_test.go:914]: Why do we need to set it here?

File: x/model/handler_test.go [hunk 12]
```diff
@@ -916,6 +934,9 @@ func TestHandler_UpdateOnlyMaxApplicableSoftwareVersion(t *testing.T) {
 
 	// try to update only max version to a value equal to stored min version
 	msgUpdateModelVersion = NewMsgUpdateModelVersion(setup.Vendor)
+
+	msgUpdateModelVersion.OtaChecksum = ""
```
Comment (ashcherbakov) [inline x/model/handler_test.go:938]: Why do we need to set it here?

File: x/model/handler_test.go [hunk 13]
```diff
@@ -975,6 +1077,9 @@ func TestHandler_OnlyOwnerAndVendorWithSameVidCanUpdateModelVersion(t *testing.T
 
 	// update existing model version by owner
 	msgUpdateModelVersion = NewMsgUpdateModelVersion(setup.Vendor)
+	msgUpdateModelVersion.OtaChecksum = ""
```
Comment (ashcherbakov) [inline x/model/handler_test.go:1080]: Why do we need to set it here?

File: x/model/handler_test.go [hunk 14]
```diff
@@ -983,6 +1088,9 @@ func TestHandler_OnlyOwnerAndVendorWithSameVidCanUpdateModelVersion(t *testing.T
 
 	// update existing model by vendor with the same VendorID as owner's one
 	msgUpdateModelVersion = NewMsgUpdateModelVersion(vendorWithSameVid)
+	msgUpdateModelVersion.OtaChecksum = ""
```
Comment (ashcherbakov) [inline x/model/handler_test.go:1091]: Why do we need to set it here?

**Review by ashcherbakov — APPROVED** [review]

**Pull request merged**